### ISSUE TITLE
Proposal: -V option to handle package variants

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#12686]
Some packages, notably CompCert and VST, have various platform variants (x86, ARM, RiscV) x (32, 64). Many VST examples work for any of these, so it is inconvenient to have different logical Coq paths for these variants - one would have to change the Require commands in the sources to switch variants.

The solution I currently follow with opam packages for CompCert and VST is to install the non-default variants in a folder Coq usually does not look at (/lib/coq-variants, a sibling folder to /lib/coq) and one uses -Q options to link one variant to VST and CompCert. The downside is that these paths are system dependent, so this does not work well, e.g. with _CoqProject.

Besides VST and CompCert there are other usages. E.g. I always dreamed about having compatible big num libraries which use either fast native integers or safe algebraic data type integers. Switching between these two would require a similar mechanism if one wants to work interactively with the fast numbers during the day and do nightly builds with the safe numbers.

My proposal is as follows: Add an option -V <folder> which has the effect that some folder below a well defined root folder, say /lib/coq-variant/ is treated as if this folder would be under /lib/coq/user-contrib.

An example: assume you have the following folder structure:
```
/lib/coq-variant/VST64/VST/...
/lib/coq-variant/VST32/VST/...
/lib/coq/user-contrib/VST/...
```
Then the option `-V VST64` should have the same effect as replacing the contents of the folder `/lib/coq/user-contrib/VST/` with the contents of the folder `/lib/coq-variant/VST64/VST/`

**Comment by Zimmi48** [comment]
First, you might not be aware of it but `$COQPATH` allows you to add any custom folder to where to search for Coq libraries. This is used all the time in the way Nix handles Coq packages. This alone should be sufficient for your use case (and preferable to `-Q` / `-R` IMHO).

Regarding the more general question that you ask on management of package variants, I think this is very related to the ongoing reflection of @ejgallego of the installation structure for version 2.0 of the Coq Dune lang. The idea is that libraries would not be made automatically available anymore but only if a user explicitly specifies that they depend on them. We could very well imagine that we would have a directory in which each library is in a separate sub-directory which is added to the loadpath only if there is an explicit dependency on this dependency. This would solve the namespace sharing / conflict issue AFAICT.

**Comment by MSoegtropIMC** [comment]
The issue is not so much building or makefiles - one can handle this.

The main issues I want to to solve with this are:

1. handle situations like _CoqProject or other GUI tool configuration. Since such paths are system dependent, putting them into _CoqProject is not portable. Currently the only portable way to handle this in CoqIDE (and likely other Coq IDEs) is to auto generate the _CoqProject file.

2. For a tool like Coq it should be very clear what things mean. I think an option like -V as proposed for package variant selection is much clearer than putting system dependent paths into an environment variable. The amount of -Q options typically used in VST projects is not easy to digest for the human eye (although this improved with the Coq platform).

I am not sure if dune mechanisms apply here. This is about interfacing different opam modules. Unless the plan is to replace opam with dune as package manager, I don't see how dune could help with such issues.

**Comment by Zimmi48** [comment]
I don't get why you think this is more related to the package manager than the build system. If you install several variants of a library with your package manager, then it could be the job of the build system to choose which version to use...

**Comment by MSoegtropIMC** [comment]
> I don't get why you think this is more related to the package manager than the build system. If you install several variants of a library with your package manager, then it could be the job of the build system to choose which version to use...

Yes, if one disables the usual library search mechanism of coq and e.g. does not map in user-contrib by default, indeed dune could handle this.

But I still want something which also works for IDEs and I want it to be clear from the coqc command lines and other configuration what this mapping is. The more configurability there is, the easier it is to get deviations between what you think you proved and what you actually proved. Right now this problem is not very prominent, but it will be as soon as there are package variants.

**Comment by Zimmi48** [comment]
I was not trying to minimize your use case but just to note the similarity with Emilio's current design considerations so that hopefully the solution to be found could address both use cases.

**Comment by MSoegtropIMC** [comment]
> I was not trying to minimize your use case but just to note the similarity with Emilio's current design considerations so that hopefully the solution to be found could address both use cases.

You are right - when we discuss a new mechanism to map modules, it should cover all requirements.

Indeed the mechanism I proposed could also make the control of modules by dune easier. Assuming one can control the root paths a simple -V module would be sufficient to enable a module. We need to discuss the exact semantics.

**Comment by gares** [comment]
My interpretation of putting an issue in a milestone is that the issue is to be addressed in order to make that release, maybe not a blocker, but still relevant. This one does not seem to fit my interpretation. Actually this issue is closer to a https://github.com/coq/ceps than an issue.

I'm removing the milestone, please react if you don't agree with this decision.

**Comment by MSoegtropIMC** [comment]
@gares : I think it is an issue which came up with the Coq platform idea to make several variants of CompCert and VST available in parallel. Afaik there is currently no reasonable (portable, user friendly, obvious, ...) solution in Coq to handle this. I definitely want to ship several versions of VST and CompCert with Coq 8.13 platform and I would see this as prerequisite unless someones has an idea how to handle this with Coq as is.

Also I would say this is relatively easy to do - if it is agreeable I can also do this.

**Comment by Zimmi48** [comment]
I think that the design is far from clear and the proposal has attracted too little feedback to conclude. IMHO:

1. this is definitely CEP material because it has multiple implications / interactions with existing tools and roadmaps of other devs;
1. therefore, I would be pretty unsatisfied with a PR being opened without a CEP having been discussed first;
1. and thus, it definitely isn't 8.13 material to me.

We are still at the step where we do not even provide a Coq Platform that would be on par with the current Windows installer. The 8.13 release of the platform should (IMHO) focus first on getting there (i.e. providing a binary installer with add-ons) before even considering all the improvements that would be worth doing. Providing variants of CompCert and VST could wait for 8.14 or could rely on non-built-in, ad hoc mechanisms for now.

**Comment by ejgallego** [comment]
The solution that works best among all the various package managers is to actually version the package itself, as it is done for libc, etc... A bit annoying, and would require that we update Coq not to finally put all of user-contrib in scope.

**Comment by MSoegtropIMC** [comment]
@ejgallego : the problem is that it is not about versions but variants. It does make a lot of sense to install say Intel + ARM and 32+64 variants of VST and CompCert in parallel in order to verify if the software actually works on all of these. This is not trivial, e.g. the definitions of what an int, a long and a long long is defers between Intel and ARM, and this obviously influences the operation of some programs.

@Zimmi48 : sure there is still a lot to do, but I don't think it would hurt to handle known problems in parallel. I must admit I don't understand why this is controversial - the proposal is I would say simple and clear, wouldn't interfere with existing mechanisms, there is no alternative proposal and the implementation would be simple. If the proposal is unclear, maybe you can tell me what is unclear about it.

**Comment by Zimmi48** [comment]
Not controversial, just not sufficiently discussed for a proposal that *does* have strong interactions with the build systems. Of course, I wouldn't oppose this feature (at all) if it had been supported by other devs, but I can only observe that nobody other than me has taken the time to comment on this issue until today (so there is an effective lack of support even if it could be for bad reasons) and I'm not the only one to think that this would deserve a CEP (Enrico said it before).

**Comment by ejgallego** [comment]
> the problem is that it is not about versions but variants. It does make a lot of sense to install say Intel + ARM and 32+64 variants of VST and CompCert in parallel in order to verify if the software actually works on all of these. This is not trivial, e.g. the definitions of what an int, a long and a long long is defers between Intel and ARM, and this obviously influences the operation of some programs.

Indeed, what I mean is that variants are captured in the package names, and they can conflict, etc... This is not ideal but may work.

Another thing to look at is Dune variants (https://dune.readthedocs.io/en/stable/variants.html) but I think they don't cover your use case.

**Comment by MSoegtropIMC** [comment]
@ejgallego : we thought about having virtual opam packages which just symlink specific variants into the default location, so that they would install and uninstall in a fraction of a second. But this would still mean to have opam calls in a multi-platform verification "build" which is not nice. This is likely what I will do until I have the -V option or a working alternative. In the end I discarded this cause of the complexities of symlinks on Windows and mixing cygwin and MinGW apps, but I think this might be better meanwhile.

Regarding dune virtual packages: how does one specify the library to use in the end if it is a system (opam) supplied library rather than a variant build by dune? I guess one can have .so files with different names? The issue with Coq is that I can't link VST into a single library and specify it as one file during a "link phase".

@Zimmi48 : still I don't see where the complexities or overlap with the build system which is worth discussing is. I would say this is fairly orthogonal to what dune might have. It is technically similar to giving different variants of a shared library with compatible interface. So while OCaml naturally has such mechanisms, Coq doesn't and this is what this issue is about.

Anyway if this is the consensus I will create a CEP.

**Comment by Zimmi48** [comment]
> I would say this is fairly orthogonal to what dune might have.

Dune (and `coq_makefile`) already need to know about some options like `-R` and `-Q` and it seems to me that it would also need to know about `-V`. Maybe I'm mistaken (I'm not the best person to discuss this) but so far I haven't been contradicted by the people in this thread that know better than me.

>It is technically similar to giving different variants of a shared library with compatible interface. So while OCaml naturally has such mechanisms

Does it? This variant feature provided by Dune is quite novel and needed a lot of refinement to work correctly. I have never heard of a similar feature that would preexist.

**Comment by ejgallego** [comment]
I think this issue is indeed a valid use case, but IMHO we should separate the use case from the implementation.

I propose we first outline a bit more declarative _what_ the needs in terms of compcert / vst, and what are the possible scenarios combinations. Then, once that is clear we can move on thinking _how_ to implement that (in the package/build system). So far it seems to me that the proposal is a bit too low-level as it depends on concrete install paths, etc...

From the build system point of view, we have been favoring a more declarative approach, that is to say, the users write a build description stating what their project provides and needs, and cannot access the details of how that is implemented. This separation has lot of advantages as otherwise you are in an scenario where users have to care about low-level details, which makes improvements and compatibility quite challenging.

**Comment by gares** [comment]
@MSoegtropIMC just swapping a .vo file under the hood does not seem to work to me. Unfortunately we don't have the equivalent of a "ABI", so if "vst.vo" was compiled against 32bit "compcert.vo", there is no way put a 64bit "compcert.vo" there and have "vst.vo" still load (even if they have the same "interface"). Is this the scenario we are discussing?

**Comment by MSoegtropIMC** [comment]
@gares: the plan is to map different sets of .vo files into a specific path and the "recompile" the files which depend on it. E.g. you map 1 64 bit ARM VST+compcert into Coq's search path and verify your C code for this and the you map a 32 bit intel VST+compcert into Coq's search path and verify your C code for that. This obviously does work. E.g. it works if I do the same with symlinks from the user-contrib folder to somewhere else.

**Comment by gares** [comment]
Oh I see, but then why can't I just have all in the Coq search path just prefixed with the arch name? Eg `Require Import x64.Compcert x32.Compcert.`, I mean changing the `-R` which is used to compile compcert, and compile it twice.

You talk about duplication, an I don't see it.

**Comment by MSoegtropIMC** [comment]
@gares: then I would have to change all the files to use the different prefix. Typically it is possible to make the verification files so that they work for any platform. So I have one set of Coq files I can compile against say 4 different implementations of some library.

Also think about the other application I mentioned: exchange a Z implementation between one using machine integers and one using inductive types. I want to use machine integers during interactive development but an axiom free solution with inductive types during nightly check runs. Obviously I don't want to change all my files then and put in different Coq paths.

**Comment by MSoegtropIMC** [comment]
P.S.: of cause I can use -Q / -R options with absolute paths. This does work but I can't put this into e.g. a _CoqProject file in a portable way. What I want is a portable way to map variants of system libraries to a Coq path, say VSTIntel32 or VSTARM64 to just VST.

**Comment by gares** [comment]
> change all my files

One last bit of info: Would you just need to change the `Require` lines, or more?

**Comment by MSoegtropIMC** [comment]
Just the Require lines - typically in form of a "From".

The point is that I might want to run several compiles against several variants in parallel. This excludes the use of symlinks in user-contrib or patching the files. I would have to either use a different opam switch or copy my complete source tree.

**Comment by gares** [comment]
My impression is still that you could compile and install all your variants and combinations thereof in `user-contrib/variant/...` and then have a `From $variant Require Import foo`. (have the value of $variant be read from _CoqProject or an env variable).

Is this totally off?

**Comment by MSoegtropIMC** [comment]
Ah OK, I didn't know I can use environment variables in a Coq file - or just in a From statement. Can I? If this does work, that would solve my problem (most likely).

**Comment by gares** [comment]
No you can't, but it seems easier to do IMO.

**Comment by MSoegtropIMC** [comment]
> No you can't, but it seems easier to implement IMO.

Possibly- at least inside a From. The -V shouldn't be that hard either I think. The largest difficulty might be to pass it through all the tools which need to be aware of it.

So you think I should close this and request environment variables in From statements?

**Comment by gares** [comment]
I personally think I need more time to think about it, and I'd like to have other's opinion on this `From $variant` proposal. I believe dune can save env vars in his build "dependencies" already. For make, well, it would not be different than any other variable set, so not a big deal either. So I see no showstoppers, but I may be wrong. We may at least discuss this at the next Coq call, next Wednesday. I've no problem integrating a trivial patch late in the release process (like the one I propose), but only if nobody foresees a problem with it.

(FTR, a CEP is just a more structured way to explain the exact needs and discuss possible designs, but it takes time... )

**Comment by ejgallego** [comment]
Indeed I think @gares is in the right direction, but even better, if I understand correctly this would be fixed the moment we stop installing all user libraries in an agglomeration in `user-contrib` but instead in each `lib/coq/$theory`, which has to be actually included in the build file as to be in scope.

**Comment by Zimmi48** [comment]
>So you think I should close this and request environment variables in From statements?

Keep it open. An issue can be about a use case rather than about a concrete solution. This is still the same use case even if the solution that is eventually implemented is different. (And we can always adapt the issue title.)

> even better, if I understand correctly this would be fixed the moment we stop installing all user libraries in an agglomeration in `user-contrib` but instead in each `lib/coq/$theory`, which has to be actually included in the build file as to be in scope.

So far, I imagined that this change would happen far off in the future, but could it actually happen earlier (maybe not 8.13, but say 8.14)? What would it take? An update to Coq, `coq_makefile`, Dune and all the custom build systems?

**Comment by gares** [comment]
My proposal is orthogonal to any change like that.

What I mean is that compcert configured for architecture a1 and a1 have to go to different logical path, and this can happen even if they are in user-contribs/. It requires the upstreams of compcert to change ./configure to pick a -R which mentions the architecture. This requires no change to any build system.

Then we could improve the syntax from `From <qualid> Require <qualid>` to `From <qualid_or_envvar> Require <qualid>` so that one can use the very same `.v` file to load `a1.x` or `a2.x`. One can also implement this outside Coq, have a `.v.in` and use `sed`, but that would be way less (non-hacker) user friendly.

FTR, Isabelle has a `~~` syntax you can use in "paths of stuff you use" that stands for the value of `$ISABELLE_HOME`.

**Comment by ejgallego** [comment]
> So far, I imagined that this change would happen far off in the future, but could it actually happen earlier (maybe not 8.13, but say 8.14)? What would it take? An update to Coq, `coq_makefile`, Dune and all the custom build systems?

Yup, basically an update to dune [already planned, guarded by version] and a new argument to coq makefile so it puts packages in scope.

> My proposal is orthogonal to any change like that.

Indeed, I think it is.

**Comment by MSoegtropIMC** [comment]
In the end this is about a structured way to put Coq libraries in scope without having to handle system dependent absolute paths by the user in makefiles, when calling CoqIDE or elsewhere.

Thinking about the env var suggestion from Enrico I still think one would need two Coq library folders:

1.) one library folder which is in scope by default (the current user-contrib)
2.) one library folder where it needs some sort of additional option to put parts of it in scope with an ability to define the logical path

If we have only Enrico's suggestion it would likely work but it might still be messy to have 4 variants of VST and/or compcert in the "default library". So we must have some way to specify the path in From without using absolute paths but also without referring to the "default library".

Regarding a dune solution: we need something which works consistently across all variants of build tools and IDEs.

I now see that this is a bit larger than I imagined - in the end we might end up in a restructuring which also requires to change a lot of opam files. So I guess it is due time to open a CEP. Will do so this weekend.

**Comment by ejgallego** [comment]
> I still think one would need two Coq library folders

Note that what we are proposing is to (finally) have each Coq package in its folder.

**Comment by ejgallego** [comment]
We did discuss about this proposal, but indeed, as @gares noted in a message above, there is the fatal problem that you cannot just replace one .vo file by another.

There is no easy solution to this problem, when you compila file that uses `foo.vo`, you cannot alter `foo.vo` and expect your file to work anymore. For example, a notation could be missing from the new version
